### PR TITLE
iox #1394 fix axivion violation for string

### PR DESCRIPTION
**Pull request by saif-at-github**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] ~~Tests follow the [best practice for testing][testing]~~
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files are added to `./clang-tidy-diff-scans.txt`
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] All touched (C/C++) source code files have been added to `./clang-tidy-diff-scans.txt`
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Relates #1394

### CODE REVIEW

**Comment by mossmaurice** [comment]
@saif-at-github There are some conflicts in `string_internal.hpp`. Could you please rebase onto `master`?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 1]
```diff
@@ -228,110 +230,110 @@ inline bool string<Capacity>::unsafe_assign(const std::string& str) noexcept
 
 template <uint64_t Capacity>
 template <typename T>
-inline IsStringOrCharArray<T, int64_t> string<Capacity>::compare(const T& other) const noexcept
+inline auto string<Capacity>::compare(const T& other) const noexcept -> IsStringOrCharArray<T, int64_t>
 {
-    uint64_t otherSize = internal::GetSize<T>::call(other);
-    auto result = memcmp(c_str(), internal::GetData<T>::call(other), std::min(m_rawstringSize, otherSize));
+    uint64_t otherSize{internal::GetSize<T>::call(other)};
+    auto result(memcmp(c_str(), internal::GetData<T>::call(other), std::min(m_rawstringSize, otherSize)));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:236]: I think here Autosar actually complains about `result` not being `const`.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 2]
```diff
@@ -535,57 +538,60 @@ string<Capacity>::insert(const uint64_t pos, const T& str, const uint64_t count)
     return true;
 }
 
+// AXIVION Next Construct AutosarC++19_03-A14.7.1:
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:541]: I couldn't find a violation of this rule. Can you please check if this suppression can be removed?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 3]
```diff
@@ -535,57 +538,60 @@ string<Capacity>::insert(const uint64_t pos, const T& str, const uint64_t count)
...
-    uint64_t length = std::min(count, m_rawstringSize - pos);
+    uint64_t length{std::min(count, m_rawstringSize - pos)};
     string subString;
     std::memcpy(&(subString.m_rawstring[0]), &m_rawstring[pos], length);
     subString.m_rawstring[length] = '\0';
     subString.m_rawstringSize = length;
     return subString;
 }
 
+// AXIVION Next Construct AutosarC++19_03-A14.7.1:
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:559]: same here

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 4]
```diff
@@ -227,7 +230,7 @@ class string
     /// @endcode
     // TruncateToCapacity_t is a compile time variable to distinguish between constructors
     // NOLINTNEXTLINE(hicpp-named-parameter, readability-named-parameter)
-    string(TruncateToCapacity_t, const char* const other) noexcept;
+    string(TruncateToCapacity_t, const char* const other) noexcept(false);
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:233]: @saif-at-github Can you please try to replace the `std::cerr` in the implementation with our `LogError()`? This may resolves the issue and you can remove the `false` again. You can find an example for the usage of `LogError` in the `SharedMemoryObject`.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 5]
```diff
@@ -580,12 +588,14 @@ class string
     /// @note out of bounds access leads to program termination
     constexpr const char& operator[](const uint64_t pos) const noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A11.3.1: It's Ok
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:591]: For the justification you could state that:
- it's a template class
- the string needs to be constructible from a string with another capacity, comparable with another string, etc.
- this doesn't break the encapsulation

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 6]
```diff
@@ -580,12 +588,14 @@ class string
...
     template <uint64_t N>
     friend class string;
 
+    // AXIVION Next Construct AutosarC++19_03-A11.3.1: It's Ok
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:595]: Here you could state that this is a factory method for the `cxx::string` and this doesn't break the encapsulation.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 7]
```diff
@@ -53,19 +55,18 @@ using IsCxxStringOrCharArray =
 
 template <typename T1, typename T2, typename ReturnType>
 using IsCxxStringOrCharArrayOrChar =
-    typename std::enable_if<(is_char_array<T1>::value || is_cxx_string<T1>::value || std::is_same<T1, char>::value)
-                                && (is_char_array<T2>::value || is_cxx_string<T2>::value
-                                    || std::is_same<T2, char>::value),
+    typename std::enable_if<((is_char_array<T1>::value || is_cxx_string<T1>::value) || std::is_same<T1, char>::value)
+                                && ((is_char_array<T2>::value || is_cxx_string<T2>::value)
+                                    || (std::is_same<T2, char>::value)),
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:60]: @FerdinandSpitzschnueffler sure.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 4 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:233]: @FerdinandSpitzschnueffler I had tried it, it doesn't resolve the issue.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 8]
```diff
@@ -500,8 +504,9 @@ class string
     ///
     /// @return true if the insertion was successful, otherwise false
     template <typename T>
-    IsCxxStringOrCharArray<T, bool> insert(const uint64_t pos, const T& str, const uint64_t count) noexcept;
+    auto insert(const uint64_t pos, const T& str, const uint64_t count) noexcept -> IsCxxStringOrCharArray<T, bool>;
 
+    // AXIVION Next Construct AutosarC++19_03-A14.7.1:
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:509]: @FerdinandSpitzschnueffler right, I see it is suppressed globally.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 9]
```diff
@@ -513,6 +518,7 @@ class string
     /// string
     optional<string<Capacity>> substr(const uint64_t pos, const uint64_t count) const noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A14.7.1:
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:521]: @FerdinandSpitzschnueffler right, I see it is suppressed globally.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 5 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:591]: @FerdinandSpitzschnueffler Sure.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 6 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:595]: @FerdinandSpitzschnueffler Sure.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 2 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:541]: @FerdinandSpitzschnueffler right, it is suppressed globally

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 3 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:559]: @FerdinandSpitzschnueffler right, it is suppressed globally

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 10]
```diff
@@ -634,14 +640,15 @@ inline constexpr char& string<Capacity>::at(const uint64_t pos) noexcept
 {
     // const_cast to avoid code duplication, safe since it's first casted to a const type and then the const is removed
     // again
+    // AXIVION Next Construct AutosarC++19_03-A5.2.3: cast char to avoid implicit conversion
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:643]: @FerdinandSpitzschnueffler Sure.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 4 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:233]: @saif-at-github I think it's because `ostream` can throw. I suggest to replace all occurrences of `std::cout` and `std::cerr` in the `cxx:string` with `LogWarn`/`LogError` and ignore the warning for now since @elBoberido's logger refactoring should not throw.

Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:233]: @MatthiasKillat Since @elBoberido is on vacation, can you maybe confirm?

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
LGTM. Once we clarified how to deal with the `noexcept` related warning, it's ready to merge IMO.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 4 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:232]: @FerdinandSpitzschnueffler @saif-at-github I think keep the `noexcept` specifier as before for now. And yes, the streams can throw since they may allocate memory, so at least `bad_alloc` is possible.

I am not sure whether the tools deduce correctly that the exception specifier is incorrect. Probably because some function not marked as `noexcept` is called (which is too weak, as it only indicates that it cannot conclude that it does not throw).

TLDR: Keep the exception specifier for now, we need to take a closer look at the rule and maybe disable it.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 11]
```diff
@@ -357,55 +359,55 @@ class string
     /// the contents of both strings are equal, an integer > 0 if the first character that does not match has a greater
     /// value in self than in other
     template <typename T>
-    IsStringOrCharArray<T, int64_t> compare(const T& other) const noexcept;
+    auto compare(const T& other) const noexcept -> IsStringOrCharArray<T, int64_t>;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:362]: @saif-at-github @FerdinandSpitzschnueffler Why is this notation preferred? It is equivalent but usually `auto` return type is used if the return type depends on compile time magic. In this case it cannot be stated after `->` explicitly but here it is additionally stated as well so the usefulness of `auto`is lost to me (it is just a more verbose way to express the same thing).

I suggest to not use this notation together with `auto` except in the case I stated above (where it is unavoidable).

Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:362]: @saif-at-github I know, but this is a purely syntactic rule and I would argue we should not follow it.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 12]
```diff
@@ -608,7 +618,7 @@ class string
 
     // safe access is guaranteed since the char array is wrapped inside the string class
     // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
-    char m_rawstring[Capacity + 1U]{'\0'};
+    char m_rawstring[Capacity + 1U]{};
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:621]: Ok, empty initializer list initializes it to `\0` as well.

@FerdinandSpitzschnueffler I am fairly certain that we can safe initialization effort by only initializing in the ctor (i.e. the actual data followed by a `\0`.). Ideally this would be the case, as 0 initialization is not free and should be avoided if not needed.

Nothing for now, but just to maybe consider in some future refactoring.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 13]
```diff
@@ -632,16 +637,17 @@ inline IsStringOrCharArray<T, optional<uint64_t>> string<Capacity>::find_last_of
 template <uint64_t Capacity>
 inline constexpr char& string<Capacity>::at(const uint64_t pos) noexcept
 {
-    // const_cast to avoid code duplication, safe since it's first casted to a const type and then the const is removed
-    // again
+    // AXIVION Next Construct AutosarC++19_03-A5.2.3: cast char to avoid implicit conversion.
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:640]:
```suggestion
    // AXIVION Next Construct AutosarC++19_03-A5.2.3: 
```

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 14]
```diff
@@ -30,20 +30,21 @@ namespace iox
 {
 namespace cxx
 {
+// AXIVION DISABLE STYLE AutosarC++19_03-A13.5.5: TODO
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:33]: I also tend to keeping the rule and moving the implementation out of class right in this pull request.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 15]
```diff
@@ -77,8 +77,8 @@ using IsCxxStringAndCxxStringOrCharArrayOrChar =
 ///     auto bar = iox::cxx::concatenate(fuu, "ahc");
 /// @endcode
 template <typename T1, typename T2>
-IsCxxStringOrCharArrayOrChar<T1, T2, string<internal::SumCapa<T1, T2>::value>> concatenate(const T1& str1,
-                                                                                           const T2& str2) noexcept;
+auto concatenate(const T1& str1, const T2& str2) noexcept
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:80]: I would like to disable the rule about trailing return syntax for template types globally. If we decide to do so, then this change and others are not necessary anymore.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 16]
```diff
@@ -138,14 +140,14 @@ class string
     /// @param [in] rhs is the copy origin
     ///
     /// @return reference to self
-    string& operator=(const string& rhs) noexcept;
+    string& operator=(const string& rhs) & noexcept;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:143]: I think we have decided to disable this rule as well, so the `&` overload is not required.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 17]
```diff
@@ -115,7 +117,7 @@ constexpr TruncateToCapacity_t TruncateToCapacity{};
 /// @brief string implementation with some adjustments in the API, because we are not allowed to throw exceptions or use
 /// heap.
 template <uint64_t Capacity>
-class string
+class string final
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:120]: I am torn about this and do not like to define classes as `final` (usually there is no need, and the STL also does not do this).

The idea appears to be to prevent a special case of inheritance that involves polymorphism when the dtor is not virtual. The downside is that it also prevents perfectly fine inheritance where this is not an issue. While we do not derive from this class, we also have no real issue with this kind of problems for multiple reasons. 

The problems require explicitly calling `delete` or the `dtor`, in the automatic case the correct dtor is called for potential derived classes.

TLDR: we can keep this `final` here but I still think the rule is unnecessary and counterproductive for useful designs.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 4 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:232]: @elBoberido @saif-at-github If I am not mistaken the plan is to not change the exception specifier anywhere for now, even if it means there are still some Axivion findings.

We will decide how to deal with this separately as it affects the API on a semantic level.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 11 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:362]: @FerdinandSpitzschnueffler @saif-at-github We should disable this rule after discussion if necessary.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 17 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:120]: @MatthiasKillat Is it not more about saying "Hey, use this class as it is. Please use composition rather than inheritance."? From a pure user-perspective I don't think there are many use-cases to inherit from classes in  `iceoryx_hoofs`.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 16 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:143]: @saif-at-github The rule is now disabled.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 4 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:232]: @FerdinandSpitzschnueffler @saif-at-github @MatthiasKillat good news everyone. With the merge of the logger this warning is gone on master :partying_face:

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 15 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:80]: Needs to be undone.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 18]
```diff
@@ -441,6 +443,7 @@ class string
     /// @brief clears the content of the string
     constexpr void clear() noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A13.5.2 , AutosarC++19_03-A13.5.3: TODO
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:446]: We need to deal with the TODO. If we cannot define it as explicit (as it breaks other ocde) we need a justification.

Justification: used for interoperability with `std::string`

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 19]
```diff
@@ -673,6 +631,60 @@ IsStdStringOrCharArrayOrChar<T, bool> operator>=(const T& lhs, const string<Capa
 /// @return the stream output of the fixed string
 template <uint64_t Capacity>
 inline std::ostream& operator<<(std::ostream& stream, const string<Capacity>& str) noexcept;
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:634]: Can you please move the above `operator<<`  above the `AXIVION DISABLE STYLE AutosarC++19_03-A13.5.5:` line? The reason is that operator does not violate the AUTOSAR rule and all the comparison operators are grouped together.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 20]
```diff
@@ -639,16 +567,17 @@ inline IsStringOrCharArray<T, optional<uint64_t>> string<Capacity>::find_last_of
...
     // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
     return const_cast<char&>(const_cast<const string<Capacity>*>(this)->at(pos));
 }
 
 template <uint64_t Capacity>
 inline constexpr const char& string<Capacity>::at(const uint64_t pos) const noexcept
 {
-    Expects(pos < size() && "Out of bounds access!");
+    // AXIVION Next Construct AutosarC++19_03-M5.3.1 : operand of logical operator is non-bool, enhances readability
+    Expects((pos < size()) && ("Out of bounds access!"));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:580]: You could replace this line with
```suggestion
    ExpectsWithMsg((pos < size()), "Out of bounds access!");
```
Then you need no suppression and the test coverage will also increase.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 21]
```diff
@@ -227,7 +229,7 @@ class string
     /// @endcode
     // TruncateToCapacity_t is a compile time variable to distinguish between constructors
     // NOLINTNEXTLINE(hicpp-named-parameter, readability-named-parameter)
-    string(TruncateToCapacity_t, const char* const other) noexcept;
+    string(TruncateToCapacity_t, const char* const other) noexcept(false);
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:234]: Please use `noexcept` everywhere without `(false)`.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 19 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:636]: Can you please explain why you reject this before resolving the conversation?

Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:636]: @FerdinandSpitzschnueffler may I have missed the comments, not intended. I'll address this on next update.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:234]: @FerdinandSpitzschnueffler   I have removed all the  `(false)` used. There will be the another violations `A15-4-2` and `A15-5-3` are reported. I'm little unclear should suppress them. Could you suggest me ?

[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:234]: @saif-at-github The other violations are all related to each other. I would propose to remove the `(false)` everywhere and ignore the related violations for now. We need to handle the problem that some functions can throw but are marked as `noexcept` probably in several places in our code base. So this could be part of a dedicated issue.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 22]
```diff
@@ -441,10 +397,13 @@ class string
     /// @brief clears the content of the string
     constexpr void clear() noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A13.5.2 , AutosarC++19_03-A13.5.3: used for interoperability with
+    // std::string
     /// @brief converts the string to a std::string
     ///
     /// @return a std::string with data equivalent to those stored in the string
-    // NOLINTNEXTLINE(hicpp-explicit-conversions) @todo iox-#260 remove this conversion and implement toStdString method
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) @ iox-#260 remove this conversion and
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:406]: Please add the `todo` again.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 23]
```diff
@@ -131,14 +132,17 @@ inline string<Capacity>::string(TruncateToCapacity_t, const char* const other, c
     }
     else if (Capacity < count)
     {
-#if defined(__GNUC__) && __GNUC__ == 8 && __GNUC_MINOR__ == 3
+// AXIVION DISABLE STYLE AutosarC++19_03-A16.0.1: conditional compilation is required for setting gcc diagnostics
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:135]: I think we should state why gcc 8 incorrectly warns here and therefore we chose the conditional compilation. On the other hand I'm not sure if maybe remove this constructor soon, see #1767 ...

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 24]
```diff
@@ -400,14 +326,16 @@ inline string<Capacity>& string<Capacity>::move(string<N>&& rhs) noexcept
...
-    uint64_t strSize = rhs.size();
+    const uint64_t strSize{rhs.size()};
     std::memcpy(&(m_rawstring[0]), rhs.c_str(), strSize);
     m_rawstring[strSize] = '\0';
     m_rawstringSize = strSize;
     rhs.clear();
     return *this;
 }
 
+// AXIVION DISABLE Style AutosarC++19_03-M5.17.1: This is not used as shift operator but as stream operator and does not require to implement '<<='
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:337]: Doesn't this warning only belong to the next construct? Why did you choose `AXIVION DISABLE Style` instead of `AXIVION Next Construct`?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 25]
```diff
@@ -400,14 +326,16 @@ inline string<Capacity>& string<Capacity>::move(string<N>&& rhs) noexcept
...
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:338]:
```suggestion
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 26]
```diff
@@ -448,7 +376,10 @@ concatenate(const T1& str1, const T2& str2, const Targs&... targs) noexcept
     return concatenate(concatenate(str1, str2), targs...);
 }
 
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:379]:
```suggestion
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 27]
```diff
@@ -664,6 +595,81 @@ inline constexpr const char& string<Capacity>::operator[](const uint64_t pos) co
...
+{
+    return (lhs.compare(rhs) > 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline IsStringOrCharArrayOrChar<T, bool> operator>=(const string<Capacity>& lhs, const T& rhs) noexcept
+{
+    return (lhs.compare(rhs) >= 0);
+}
+// AXIVION ENABLE Style AutosarC++19_03-M5.17.1
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:672]: Shouldn't this be
```suggestion
// AXIVION ENABLE Style AutosarC++19_03-A13.5.5
```
?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 23 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:135]: @FerdinandSpitzschnueffler should I keep this suppression for now? once the c'tor removed this may revisit, based on this #1767

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 24 above)
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:337]: @FerdinandSpitzschnueffler `M5.17.1` violation has been reported multiple place in the file so decided to use the `DISABLE Style` instead of `AXIVION Next Construct` let me check and  may use the  `AXIVION Next Construct`

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 23 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:137]: Since the discussion is still ongoing and I'm quite sure that we will forget to revisit if it will be decided to keep the constructor, I would propose to extend the justification to
``` conditional compilation is required for setting gcc diagnostics since ggc 8 incorrectly warns here about out of bounds array access```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 28]
```diff
@@ -23,6 +23,7 @@ namespace iox
 {
 namespace cxx
 {
+// AXIVION DISABLE STYLE AutosarC++19_03-A12.6.1 : false positive, field intentionally not initialized
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:26]: @FerdinandSpitzschnueffler this was reported in my local machine for `m_rawstring` but not in dashboad, I shall remove it.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 29]
```diff
@@ -442,10 +398,13 @@ class string
...
     constexpr void clear() noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A13.5.2 , AutosarC++19_03-A13.5.3: used for interoperability with
+    // std::string
     /// @brief converts the string to a std::string
     ///
     /// @return a std::string with data equivalent to those stored in the string
-    // NOLINTNEXTLINE(hicpp-explicit-conversions) @todo iox-#260 remove this conversion and implement toStdString method
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) @todo iox-#260 remove this conversion and
+    // implement toStdString method
```
[the remark this answers is not in the document]
Comment (saif-at-github) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:407]: @FerdinandSpitzschnueffler I see that, corrected now.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]
